### ISSUE TITLE
Evaluation of HF Models with LIDA

### ISSUE THREAD

**Issue by victordibia** [microsoft/lida#41]
## What 

Local models (e.g. LLAMA based models available via HuggingFace in the 7B or 13B size classes) offer multiple benefits (e.g., can be finetuned/adapted, run locally etc).
While LIDA has been mostly tested with OpenAI models, more work is needed to test workflows and performance for HF models.

## Work Items 

- Test a set of local HF models either directly with llmx or with LIDA, systematically document bugs and suggest fixes via PRs

**Comment by trojrobert** [comment]
I have started working on this

**Comment by victordibia** [comment]
Great. Let us discuss your findings here so far.  

On my end, I have been trying local huggingface models and have added the following to the documentation readme. 

For example, I have found this [hermes 13b](https://huggingface.co/uukuguy/speechless-llama2-hermes-orca-platypus-13b) to decent performance in terms of goal generation but limited success with visualization generation. 

I am thinking of creating a seperate repo for experiments using lida to generate visualizations.

-----

## Using LIDA with Locally Hosted LLMs (HuggingFace)

LIDA uses the [llmx](https://github.com/victordibia/llmx) library as its interface for text generation. llmx supports multiple local models including HuggingFace models. You can use the huggingface models directly (assuming you have a gpu) or connect to an openai compatible local model endpoint e.g. using the excellent [vllm](https://vllm.readthedocs.io/en/latest/) library.

### Using HuggingFace Models Directly

```python
from lida import llm
text_gen = llm(provider="hf", model="uukuguy/speechless-llama2-hermes-orca-platypus-13b", device_map="auto")
lida = Manager(llm=text_gen)
# now you can call lida methods as above e.g.
sumamry = lida.summarize("data/cars.csv") # ....
```

### Using an OpenAI Compatible Endpoint e.g. [vllm server](https://vllm.readthedocs.io/en/latest/getting_started/quickstart.html#openai-compatible-server)

```python
from lida import Manager, TextGenerationConfig , llm

model_name = "uukuguy/speechless-llama2-hermes-orca-platypus-13b"
model_details = [{'name': model_name, 'max_tokens': 2596, 'model': {'provider': 'openai', 'parameters': {'model': model_name}}}]

# assuming your vllm endpoint is running on localhost:8000
text_gen = llm(provider="openai",  api_base="http://localhost:8000/v1", api_key="EMPTY", models=model_details)
lida = Manager(text_gen = text_gen)
```

**Comment by trojrobert** [comment]
Yes, I was able to recommend we create a separate branch for this

**Comment by zirubak** [comment]
Do you guys consider using https://mistral.ai/news/announcing-mistral-7b/ ?

**Comment by trojrobert** [comment]
@victordibia I think the code below in llmx library cause this error. Line 47 and 48. You are adding "provider" and "models" to the kwargs and you also have it as an argument. 

```
    kwargs["provider"] = kwargs["provider"] if "provider" in kwargs else provider
    kwargs["models"] = kwargs["models"] if "models" in kwargs else models
```

![image](https://github.com/microsoft/lida/assets/28733978/6e0d02ad-e8b9-44c2-8293-797fe3adeee5)

**Comment by trojrobert** [comment]
@victordibia I got an error while downloading other models from hugging face

This is the code
![image](https://github.com/microsoft/lida/assets/28733978/ce182f63-6dad-4e10-af6e-9983af9f765f)

This is the error
![image](https://github.com/microsoft/lida/assets/28733978/f46e1643-3d57-4d01-89ec-f5a08b96cf45)

Solution

Add new argument "offload_folder" to self.model in llmx package
![image](https://github.com/microsoft/lida/assets/28733978/05f15b18-af63-46fa-b8d9-35494494c59d)

**Comment by vegansquirrel** [comment]
Hello, I am having some trouble Loading "ehartford/dolphin-2.5-mixtral-8x7b"
Has anyone tried this? 
Any help or steps you can provide is much appreciated.

**Comment by victordibia** [comment]
Hi,

I have not tested with the mixtral model series.
I'd suggest attempting to use [vllm](https://github.com/vllm-project/vllm) to setup an open ai compatible server and then connect to that using the openai approach. Mixtral is supported on vllm.

Let me know how it goes.

**Comment by kishoretvk** [comment]
> Do you guys consider using https://mistral.ai/news/announcing-mistral-7b/ ?

this works for summary generation not for graphs.
did you find anything else working better ?
